source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -169,52 +174,46 @@ void FilterConfigSubscription::onConfigUpdateFailed(Config::ConfigUpdateFailureR
-    Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix) {
-  // FilterConfigSubscriptions are unique based on their config source and filter config name
-  // combination.
-  // TODO(https://github.com/envoyproxy/envoy/issues/11967) Hash collision can cause subscription
-  // aliasing.
-  const std::string subscription_id = absl::StrCat(MessageUtil::hash(config_source), ".", name);
-  auto it = subscriptions_.find(subscription_id);
-  if (it == subscriptions_.end()) {
-    auto subscription = std::make_shared<FilterConfigSubscription>(
-        config_source, name, factory_context, stat_prefix, *this, subscription_id);
-    subscriptions_.insert({subscription_id, std::weak_ptr<FilterConfigSubscription>(subscription)});
-    return subscription;
-  } else {
-    auto existing = it->second.lock();
-    ASSERT(existing != nullptr,
-           absl::StrCat("Cannot find subscribed filter config resource ", name));
-    return existing;
-  }
+FilterConfigProviderManagerImpl::~FilterConfigProviderManagerImpl() {
+  // assert that all providers have been destroyed.
+  ASSERT(dynamic_filter_config_providers_.empty());
 }
 
 FilterConfigProviderPtr FilterConfigProviderManagerImpl::createDynamicFilterConfigProvider(
     const envoy::config::core::v3::ConfigSource& config_source,
     const std::string& filter_config_name, const std::set<std::string>& require_type_urls,
     Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
     bool apply_without_warming) {
-  auto subscription =
-      getSubscription(config_source, filter_config_name, factory_context, stat_prefix);
-  // For warming, wait until the subscription receives the first response to indicate readiness.
-  // Otherwise, mark ready immediately and start the subscription on initialization. A default
-  // config is expected in the latter case.
-  if (!apply_without_warming) {
-    factory_context.initManager().add(subscription->initTarget());
-  }
-  auto provider = std::make_unique<DynamicFilterConfigProviderImpl>(
-      std::move(subscription), require_type_urls, factory_context);
-  // Ensure the subscription starts if it has not already.
-  if (apply_without_warming) {
-    factory_context.initManager().add(provider->init_target_);
+  const std::string provider_identifier =
+      absl::StrCat(MessageUtil::hash(config_source), ".", filter_config_name);
+  auto it = dynamic_filter_config_providers_.find(provider_identifier);
+  if (it == dynamic_filter_config_providers_.end()) {
+    auto subscription = std::make_shared<FilterConfigSubscription>(
+        config_source, filter_config_name, factory_context, stat_prefix);
+    // For warming, wait until the subscription receives the first response to indicate readiness.
+    // Otherwise, mark ready immediately and start the subscription on initialization. A default
+    // config is expected in the latter case.
```

Comment:
Disclaimer: I haven't understood the entire change. It seems to me that subscriptions and filter config providers cannot be 1-1 since filter config provider may have different type_urls and warming settings. It's possible to have two providers that have different defaults, but subscribe to the same resource. How do we handle that here?

Reply:
Thinking out loud: maybe we need to initialize unique config providers better from the same subscription? That way we can handle "do not warm if already sent" and "duplicate config pushed" cases better.

Reply:
> It seems to me that subscriptions and filter config providers cannot be 1-1 since filter config provider may have different type_urls and warming settings. It's possible to have two providers that have different defaults, but subscribe to the same resource. How do we handle that here?

I thought about this a bit. `apply_without_warming` was handled, by adding subscription init target into init manager if `apply_without_warming=false`. Default config does not seem to be a concern for provider. It is dealt within `HttpConnectionManagerConfig`, one level above provider. I did not consider `require_type_urls`. If we are going with change as this PR, we probably could add it into provider id hash?

> maybe we need to initialize unique config providers better from the same subscription? That way we can handle "do not warm if already sent" and "duplicate config pushed" cases better.

Are you thinking about something like https://github.com/envoyproxy/envoy/pull/14985/commits/6c677a48702a966f6333ec689198f631892f8310 ?